### ISSUE TITLE
KVM Template Clone: VM Config (specs) not amended post-clone to = Module Plan

### ISSUE THREAD

**Issue by pqhostde** [The-Network-Crew/Proxmox-VE-for-WHMCS#32]
I was following the pdf instruction, now i get: 
 
PVE2 API: This API Request Failed. HTTP CODE: 400, HTTP ERROR: HTTP/1.1 400 Parameter verification failed., REPLY INFO: {"errors":{"newid":"invalid format - value does not look like a valid VM ID\n"},"data":null}**

**Comment by pqhostde** [comment]
Thanks, you are right, the test System was a clean WHMCS.
So we have another Problem, when creating VM it works fine, but the Ressource wound be set from the VM plan.

**Comment by lsthompson** [comment]
>  So we have another Problem, when creating VM it works fine, but the Ressource wound be set from the VM plan.

Are you able to elaborate on this please? I'm not sure that I'm understanding. Thanks!

**Comment by pqhostde** [comment]
> > So we have another Problem, when creating VM it works fine, but the Ressource wound be set from the VM plan.
> 
> Are you able to elaborate on this please? I'm not sure that I'm understanding. Thanks!

the problem seems to be a missing feature, if I clone a template there is no adjustment afterwards via cloudinit support.
no cpu or RAM or network settings are made after the clone process. This works via ISO, but this would also be very helpful with a template.

**Comment by lsthompson** [comment]
> If I clone a template there is no adjustment afterwards via cloudinit support.
> No CPU or RAM or network settings are made after the clone process.

Thank you. Can you please elaborate on what should be happening vs what is happening?

You mentioned that the instance gets specs from the Module Plan, which is what we would expect.

Are you trying to then change the specs after the fact via cloud-init? A stepped-out process would be good.

**Comment by lsthompson** [comment]
@pqhostde - this has some time in our current dev sprint, do you have time to assist here? Thank you!

**Comment by jbrouhard98** [comment]
Cloning a vm in proxmox will clone the originals hardware setup identically.   So what should ideally happen is a clone is done and before startup the hardware should be modified.    Perhaps this just needs an extra step?

**Comment by lsthompson** [comment]
That makes sense, though you still have the issue of disk expansion. There isn't just the PVE-level expansion, there's everything within the virtualised instance - depending on set-up this includes PV/LV/VG/partition.

However, it sounds like this isn't functioning as intended at the moment. Checking the code ([L90](https://github.com/The-Network-Crew/Proxmox-VE-for-WHMCS/blob/master/modules/servers/pvewhmcs/pvewhmcs.php#L90)-124) though, it seems a KVM Template clone is quite basic. It makes a full clone, and uses the WHMCS Service ID as the PVE "newid".

It fits, as you can't declare new specs: https://pve.proxmox.com/pve-docs/api-viewer/#/nodes/{node}/qemu/{vmid}/clone

So I think you're on the ball with us needing to step this out into clone-then-respec. I don't think that's cloud-init territory.

> So what should ideally happen is a clone is done and before startup the hardware should be modified. Extra step.

**Comment by lsthompson** [comment]
@jbrouhard98 @pqhostde

Could you please give this a whirl now? See if it makes the changes.

**Comment by pqhostde** [comment]
> @jbrouhard98 @pqhostde
> 
> Could you please give this a whirl now? See if it makes the changes.

Hello,

i will spin up a lab and Test it.
Give me a few days.

Best regards.

**Comment by lsthompson** [comment]
@pqhostde how are you going with your testing?

We would like to release this fine-tuning release.

**Comment by lsthompson** [comment]
@jbrouhard98 are you able to please help with testing mate?

Otherwise I have to wait for the new dev machine, then run through templating config.

**Comment by jbrouhard98** [comment]
I can try testing this next week when I can spool up a new test environment.   Been a crazy few days here

**Comment by lsthompson** [comment]
Race of the test-beds! Thank you kindly good sir, appreciate it in advance if you find a moment

**Comment by pqhostde** [comment]
> @pqhostde how are you going with your testing?
> 
> We would like to release this fine-tuning release.
I was now able to test, the VM is being cloned but unfortunately, I still don't see any change, the hard disk has not been changed and the CPU, RAM and IP settings in Cloudinit have not been set either. I tried to create a KVM virtual machine.

**Comment by lsthompson** [comment]
Thank you @pqhostde - do you see any other messages on the cluster?

Is the VM ID in that message for the source or target instance? ie. to-clone or cloned?

> the hard disk has not been changed and the CPU, RAM and IP settings in Cloudinit have not been set either

We won't do storage as it's multi-stage and changes needed in-OS too. Happy to reconsider via PR.

You mention Cloudinit, so where are you hoping that the resources will be obtained from exactly?

As primarily the Module is concerned with the VM/CT Plans defined within the Module Config.

**Comment by pqhostde** [comment]
By Cloudinit I meant the IP settings and password. If you have a base template and expand the VM-Disk-space before starting the Cloudinit template, it will automatically increase when the VM boots up. That's why Cloudinit support wouldn't be so good for KVM Provisioning. But regardless of the cloud init, the cloned VM should have at least expanded resources such as CPU, RAM and memory and unfortunately this does not happen, and no message appears in the cluster that the VM has been updated. What debug information can I provide so that this function works properly?

> Thank you @pqhostde - do you see any other messages on the cluster?
> 
> Is the VM ID in that message for the source or target instance? ie. to-clone or cloned?
> 
> > the hard disk has not been changed and the CPU, RAM and IP settings in Cloudinit have not been set either
> 
> We won't do storage as it's multi-stage and changes needed in-OS too. Happy to reconsider via PR.
> 
> You mention Cloudinit, so where are you hoping that the resources will be obtained from exactly?
> 
> As primarily the Module is concerned with the VM/CT Plans defined within the Module Config.

**Comment by jbrouhard98** [comment]
This should happen after clone is complete.  It won't happen during progress as you're skipping a step

**Comment by lsthompson** [comment]
> By Cloudinit I meant the IP settings and password.

It should honour what's configured in your cluster. If in doubt, get PVE Enterprise Subscription for Support.

Otherwise, there is always the Proxmox Forum which has a huge number of users who you can approach. :)

> If you have a base template and expand the VM-Disk-space before starting the Cloudinit template, it will automatically increase when the VM boots up.

It will do part of the process, as Joe talks to as well. In an LVM-based storage architecture, yes the PVE-level increase can be completed, though you still have the VG/LV/etc and partitioning to expand/adjust to suit the new-found space. 

Unless you have a Pull Request which handles all of that variable in-OS somewhat-package-dependent work, then we don't want to get into doing 50% of a job!

> the cloned VM should have at least expanded resources such as CPU, RAM etc and unfortunately this does not happen
> 
> no message appears in the cluster that the VM has been updated. What debug information can I provide

Agreed. Hmm, I just tacked on some amendment lines after the clone fires. 

@jbrouhard98 did you find time to test?

**Comment by pqhostde** [comment]
So I tested, But I can imagine using the cluster resource scheduler. That the request is executed on a different node and therefore the adjustment does not take place. I'll probably have to wait for version 1.3 until clusters are supported.

> > the cloned VM should have at least expanded resources such as CPU, RAM etc and unfortunately this does not happen
> > no message appears in the cluster that the VM has been updated. What debug information can I provide
> 
> Agreed. Hmm, I just tacked on some amendment lines after the clone fires.
> 
> @jbrouhard98 did you find time to test?

**Comment by jbrouhard98** [comment]
> 
> @jbrouhard98 did you find time to test?

Afraid not.  Currently on a time crunch getting a new client set up.   

We also kind of ditched whmcs after a billing snafu.   I'll see about restoring my dev setup after the new year.  Best I can offer at this time :(

**Comment by lsthompson** [comment]
All good. Our new test bed will be live before the end of the year.

**Comment by lsthompson** [comment]
> I'll see about restoring my dev setup after the new year. Best I can offer at this time :(

@jbrouhard98 do you have some time now? We also have a test-bed online.

**Comment by lsthompson** [comment]
@pqhostde any difference now if you update (and SQL) from repository itself?

ie. So you get this: https://github.com/The-Network-Crew/Proxmox-VE-for-WHMCS/commit/ddbc25b1f57e36a37777f7c3be9934ab5240f7db

**Comment by pqhostde** [comment]
Hello,

Unfortunately, I no longer have a WHMCS license because we changed the solution.

___________________________________
Mit freundlichen Grüßen / Best Regards

Maebou Ngum
Geschäftsleitung • PQHost UG
Telefon: 05404 9939999
Email: ***@***.******@***.***>

www.pqhost.net<http://www.pqhost.net/>
<https://aka.ms/AAb9ysg>
________________________________
